### ISSUE TITLE
poetry startup time is slow

### ISSUE THREAD

**Issue by yoav-orca** [python-poetry/poetry#3502]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: macOS cataline 
- **Poetry version**: 1.1.4
- **The contents of your pyproject.toml file**:
```
[tool.poetry]
name = "poetry_slow_startup"
version = "0.1.0"
description = ""
authors = ["John Doe <user@acme.com>"]

[tool.poetry.dependencies]
python = "^3.9"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

## Issue
I'll start by saying that poetry is awesome, it made me and my team more productive since we migrated from Pipenv!

I use the poetry `run`/`shell` commands a lot in my daily routine, I also use poetry as a wrapper to other tools that I use from my IDE & git hooks. However, there's an overhead when running commands with poetry:
```bash
$ time /usr/bin/true
/usr/bin/true  0.00s user 0.00s system 69% cpu 0.003 total
$ time /usr/bin/true
/usr/bin/true  0.00s user 0.00s system 72% cpu 0.006 total
$ time /usr/bin/true
/usr/bin/true  0.00s user 0.00s system 78% cpu 0.005 total

# now with poetry

$ time poetry run /usr/bin/true
poetry run /usr/bin/true  0.54s user 0.15s system 72% cpu 0.944 total
$ time poetry run /usr/bin/true
poetry run /usr/bin/true  0.48s user 0.09s system 98% cpu 0.581 total
$ time poetry run /usr/bin/true
poetry run /usr/bin/true  0.52s user 0.14s system 78% cpu 0.844 total
```

It seems that running commands with poetry adds between 570-930ms latency to the command.

I tried to profile the issue, and it seems that most of the time is spent in the import statements before the main is called, the main itself is only 90ms.
I tried to profile the imports with `PYTHONPROFILEIMPORTTIME` as follows:
```shell
PYTHONPROFILEIMPORTTIME=1 poetry run /usr/bin/true 2>imports.txt
```

I visualized the results using [tuna](https://github.com/nschloe/tuna) as follows:
```shell
tuna imports.txt
```

![image](https://user-images.githubusercontent.com/65133955/102699396-00e73700-424d-11eb-9d4b-0b440f456ffc.png)

The results indicate that most of the time is used in imports that are not used by the `run`/`shell` commands. 
I propose using [LazyLoader](https://docs.python.org/3/library/importlib.html#importlib.util.LazyLoader) to improve the startup time and responsiveness of poetry.

**Comment by sinoroc** [comment]
Probably could have been filed as "feature request" instead of "bug".

**Comment by yoav-orca** [comment]
@sinoroc I wasn't sure, because it didn't feel like a feature request per se, because older versions of poetry were faster (poetry 1.0 is consistently under 350ms), but I can change the label

**Comment by sinoroc** [comment]
> @sinoroc I wasn't sure, because it didn't feel like a feature request per se, because older versions of poetry were faster (poetry 1.0 is consistently under 350ms), but I can change the label

No worries. I just mentioned it. Maintainers will decide.

Maybe it is worth editing the first message to clearly show that it could be a regression (i.e. it used to be faster).

**Comment by jules-ch** [comment]
A bunch of imports for commands need to be moved to `handle` to delay imports.
It's done like that for some commands & imports but not all.

**Comment by yoav-orca** [comment]
@jules-ch could you point me to specific candidates, I'll try to send a PR

**Comment by jules-ch** [comment]
#3618 might help. I think we can wait & see if that improves performance.

**Comment by yoav-orca** [comment]
@jules-ch I gave poetry 1.1.5 a try, it did slightly improve things, but it's still about ~500ms slower than native code run.
I re-run tuna to see where the time is spent on imports, the situation is not very different, most of the time is spent before the main logic happens.

![image](https://user-images.githubusercontent.com/65133955/114578558-a8b5c400-9c85-11eb-907c-a75649787c66.png)

**Comment by abn** [comment]
@yoav-orca can you try master? Should get you to around ~300ms. And #3938 will get you to ~250ms.

### Using pipx
```sh
pipx install --suffix=@3938 'poetry @ git+https://github.com/python-poetry/poetry.git@refs/pull/3938/head'
PYTHONPROFILEIMPORTTIME=1 poetry@3938 run /usr/bin/true 2>imports.txt 
```

```sh
pipx install --suffix=@3938 'poetry @ git+https://github.com/python-poetry/poetry.git@refs/pull/3938/head'
PYTHONPROFILEIMPORTTIME=1 poetry@3938 run /usr/bin/true 2>imports.txt 
```

**Comment by yoav-orca** [comment]
Hi @abn I tried both versions

My setup:
macOS Catalina
Python 3.9.4

Empty poetry 
```
mkdir test-proj
cd test-proj
poetry init -n
```

With 3938:
```
» time poetry@3938 run /usr/bin/true
poetry@3938 run /usr/bin/true  0.31s user 0.07s system 97% cpu 0.391 total
» time poetry@3938 run /usr/bin/true
poetry@3938 run /usr/bin/true  0.32s user 0.08s system 97% cpu 0.403 total
» time poetry@3938 run /usr/bin/true
poetry@3938 run /usr/bin/true  0.31s user 0.07s system 97% cpu 0.399 total
```
Tuna on the dependencies

At master:
```
» time poetry@master run /usr/bin/true
poetry@master run /usr/bin/true  0.31s user 0.07s system 97% cpu 0.394 total
» time poetry@master run /usr/bin/true
poetry@master run /usr/bin/true  0.32s user 0.07s system 97% cpu 0.405 total
» time poetry@master run /usr/bin/true
poetry@master run /usr/bin/true  0.31s user 0.07s system 97% cpu 0.396 total
```
The diagram looks very similar to both branches, I'm adding the diagram for 3938:
![image](https://user-images.githubusercontent.com/65133955/115138762-b6c56500-a036-11eb-9219-70d3384707b1.png)

BTW when running on a non-empty project (~50 dependencies) I see a spike of 100-150 ms on startup time.

**Comment by lafrenierejm** [comment]
May I ask what the target performance is for this issue to be considered resolved?

Based on [yoav-orca's latest comment](https://github.com/python-poetry/poetry/issues/3502#issuecomment-821952685), it looks like the timing of recent `master` is now ~350ms and similar to 1.0's timing:
> older versions of poetry were faster (poetry 1.0 is consistently under 350ms)

**Comment by systematicguy** [comment]
We really need this to get going/resolved as it is the sole serious issue in the 1.2 project besides another one.

**Comment by neersighted** [comment]
I'm going to close this as current speeds are not a regression -- I'd suggest opening issues for specific commands/sections of code with specific targets if anyone finds any unnecessary hot loops or blocking behavior.
